### PR TITLE
Feature/sa 51 wishlists service

### PR DESCRIPTION
**Pull request by cuqmbr**

### CODE REVIEW

File: ShoppingAssistantApi.Domain/Entities/Wishlist.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+namespace ShoppingAssistantApi.Domain.Entities;
+
+public class Wishlist : EntityBase
+{
+    public required string Name { get; set; }
+    public required string Type { get; set; }
+    public ICollection<Message>? Messages { get; set; } = null;
+
+    public required ObjectId UserId { get; set; }
```
Comment (Mikhail-Beloded) [inline ShoppingAssistantApi.Domain/Entities/Wishlist.cs]: You don't need this. You already have property for user as CreatedById from EntityBase

File: ShoppingAssistantApi.Domain/Entities/Message.cs [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+using ShoppingAssistantApi.Domain.Common;
+
+namespace ShoppingAssistantApi.Domain.Entities;
+
+public class Message : EntityBase
+{
+    public required string Text { get; set; }
+    public required string Role { get; set; }
+
+    public ObjectId? WishlistId { get; set; } = null;
```
Comment (Mikhail-Beloded) [inline ShoppingAssistantApi.Domain/Entities/Message.cs]: In ER diagram it's not null

File: ShoppingAssistantApi.Domain/Enums/MessageRoles.cs [hunk 3]
```diff
@@ -0,0 +1,7 @@
+namespace ShoppingAssistantApi.Domain.Enums;
+
+public enum MessageRoles
+{
+    User = 0,
```
Comment (Mikhail-Beloded) [inline ShoppingAssistantApi.Domain/Enums/MessageRoles.cs:5]: The same values of constants (0 and 0)

File: ShoppingAssistantApi.Domain/Entities/Wishlist.cs [hunk 4]
```diff
@@ -0,0 +1,13 @@
+using MongoDB.Bson;
+using ShoppingAssistantApi.Domain.Common;
+
+namespace ShoppingAssistantApi.Domain.Entities;
+
+public class Wishlist : EntityBase
+{
+    public required string Name { get; set; }
```
Comment (Mikhail-Beloded) [inline ShoppingAssistantApi.Domain/Entities/Wishlist.cs:7]: Add spaces between properties

File: ShoppingAssistantApi.Domain/Entities/Message.cs [hunk 5]
```diff
@@ -0,0 +1,12 @@
+using MongoDB.Bson;
+using ShoppingAssistantApi.Domain.Common;
+
+namespace ShoppingAssistantApi.Domain.Entities;
+
+public class Message : EntityBase
+{
+    public required string Text { get; set; }
+    public required string Role { get; set; }
```
Comment (Mikhail-Beloded) [inline ShoppingAssistantApi.Domain/Entities/Message.cs:10]: Add spaces between properties

File: ShoppingAssistantApi.Domain/Entities/Wishlist.cs [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+    public required string Type { get; set; }
+    public ICollection<Message>? Messages { get; set; } = null;
```
Comment (Mikhail-Beloded) [inline ShoppingAssistantApi.Domain/Entities/Wishlist.cs]: = null part unnecessary

File: ShoppingAssistantApi.Domain/Entities/Message.cs (hunk 2 above)
Comment (Mikhail-Beloded) [inline ShoppingAssistantApi.Domain/Entities/Message.cs]: = null part unnecessary

File: ShoppingAssistantApi.Api/Mutations/WishlistsMutation.cs
Comment (Mikhail-Beloded) [inline ShoppingAssistantApi.Api/Mutations/WishlistsMutation.cs:1]: What about to create query for retrieving data such as getting wishlists from database?

File: ShoppingAssistantApi.Application/Models/CreateDtos/WishlistCreateDto.cs [hunk 7]
```diff
@@ -0,0 +1,7 @@
+namespace ShoppingAssistantApi.Application.Models.CreateDtos;
+
+public class WishlistCreateDto
+{
+    public required string Type { get; set; }
+    public required string FirstMessageText { get; set; }
```
Comment (Mikhail-Beloded) [inline ShoppingAssistantApi.Application/Models/CreateDtos/WishlistCreateDto.cs:7]: spacing again

File: ShoppingAssistantApi.Application/Models/Dtos/MessageDto.cs [hunk 8]
```diff
@@ -0,0 +1,11 @@
+namespace ShoppingAssistantApi.Application.Models.Dtos;
+
+public class MessageDto
+{
+    public required string Id { get; set; }
+
+    public required string Text { get; set; }
```
Comment (Mikhail-Beloded) [inline ShoppingAssistantApi.Application/Models/Dtos/MessageDto.cs:7]: spacing again

File: ShoppingAssistantApi.Application/Models/Dtos/WishlistDto.cs [hunk 9]
```diff
@@ -0,0 +1,11 @@
+namespace ShoppingAssistantApi.Application.Models.Dtos;
+
+public class WishlistDto
+{
+    public required string Id { get; set; }
+
+    public required string Name { get; set; }
+    public required string Type { get; set; }
```
Comment (Mikhail-Beloded) [inline ShoppingAssistantApi.Application/Models/Dtos/WishlistDto.cs:9]: spacing again

File: ShoppingAssistantApi.Infrastructure/Services/WishlistsService.cs [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+using ShoppingAssistantApi.Application.Paging;
+using ShoppingAssistantApi.Domain.Entities;
+using ShoppingAssistantApi.Domain.Enums;
+
+namespace ShoppingAssistantApi.Infrastructure.Services;
+
+public class WishlistsService : IWishlistsService
+{
+    private readonly IWishlistsRepository _wishlistsRepository;
+    private readonly IMessagesRepository _messagesRepository;
```
Comment (Mikhail-Beloded) [inline ShoppingAssistantApi.Infrastructure/Services/WishlistsService.cs:19]: spacing

File: ShoppingAssistantApi.Infrastructure/Services/WishlistsService.cs [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+        _mapper = mapper;
+    }
+
+    public async Task<WishlistDto> StartPersonalWishlistAsync(WishlistCreateDto dto, CancellationToken cancellationToken)
+    {
+        var newWishlist = _mapper.Map<Wishlist>(dto);
+
+        newWishlist.CreatedById = (ObjectId) GlobalUser.Id;
+        newWishlist.CreatedDateUtc = DateTime.UtcNow;
+        newWishlist.Name = $"{newWishlist.Type} Search";
```
Comment (Mikhail-Beloded) [inline ShoppingAssistantApi.Infrastructure/Services/WishlistsService.cs:42]: In this way you will have wishlists only with 2 types of names

File: ShoppingAssistantApi.Infrastructure/Services/WishlistsService.cs [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+        if (!ObjectId.TryParse(wishlistId, out var wishlistObjectId))
+        {
+            throw new InvalidDataException("Provided id is invalid.");
+        }
+        newMessage.WishlistId = wishlistObjectId;
+        newMessage.Role = MessageRoles.User.ToString();
+        newMessage.CreatedById = (ObjectId) GlobalUser.Id;
+        newMessage.CreatedDateUtc = DateTime.UtcNow;
+
+        var relatedWishlistPage = await _wishlistsRepository.GetPageAsync(1, 1, x => x.Id == wishlistObjectId && x.CreatedById == GlobalUser.Id, cancellationToken);
```
Comment (Mikhail-Beloded) [inline ShoppingAssistantApi.Infrastructure/Services/WishlistsService.cs]: avoid this and create in WishListRepository GetWishList method for getting wishlists and put expression as parameter to make that method more flexible

File: ShoppingAssistantApi.Persistance/PersistanceExtentions/DbInitialaizer.cs [hunk 13]
```diff
@@ -1,41 +1,42 @@
...
+using ShoppingAssistantApi.Domain.Enums;
+using ShoppingAssistantApi.Persistance.Database;
 
 namespace ShoppingAssistantApi.Persistance.PersistanceExtentions;
 
 public class DbInitialaizer
 {
     private readonly IUsersService _usersService;
-
     private readonly IUserManager _userManager;
```
Comment (Mikhail-Beloded) [inline ShoppingAssistantApi.Persistance/PersistanceExtentions/DbInitialaizer.cs]: bring spacing back

File: ShoppingAssistantApi.Tests/Tests/WishlistsTests.cs
Comment (Mikhail-Beloded) [inline ShoppingAssistantApi.Tests/Tests/WishlistsTests.cs:1]: split your testing method for both mutations

Comment (Mikhail-Beloded) [inline ShoppingAssistantApi.Tests/Tests/WishlistsTests.cs:1]: Also add testing methods with invalid data

File: ShoppingAssistantApi.Api/ApiExtentions/GraphQlExtention.cs [hunk 14]
```diff
@@ -16,6 +16,7 @@ public static IServiceCollection AddGraphQl(this IServiceCollection services)
                 .AddTypeExtension<AccessMutation>()
```
Comment (cuqmbr) [inline ShoppingAssistantApi.Api/ApiExtentions/GraphQlExtention.cs:16]: test comment

**Review by Mikhail-Beloded — CHANGES_REQUESTED** [review]

File: ShoppingAssistantApi.Api/Mutations/WishlistsMutation.cs
Comment (cuqmbr) [reply ShoppingAssistantApi.Api/Mutations/WishlistsMutation.cs:1]: will be added in [SA-53](https://shchoholiev.atlassian.net/browse/SA-53?atlOrigin=eyJpIjoiZjc2YzVjZWVkMDU0NGFkYjk3OTRlMmU4NzQ0Mzc5YTUiLCJwIjoiaiJ9) this week

File: ShoppingAssistantApi.Infrastructure/Services/WishlistsService.cs (hunk 11 above)
Comment (cuqmbr) [reply ShoppingAssistantApi.Infrastructure/Services/WishlistsService.cs:42]: Yes, by default. Name will be changed later automatically based on user messages or manually by the user.

File: ShoppingAssistantApi.Api/Queries/WishlistsQuery.cs [hunk 15]
```diff
@@ -0,0 +1,15 @@
+using HotChocolate.Authorization;
+using ShoppingAssistantApi.Application.IServices;
+using ShoppingAssistantApi.Application.Models.Dtos;
+using ShoppingAssistantApi.Application.Paging;
+
+namespace ShoppingAssistantApi.Api.Queries;
+
```
Comment (Mikhail-Beloded) [inline ShoppingAssistantApi.Api/Queries/WishlistsQuery.cs:7]: Add here request for retrieving single personal wishlist. GetPersonalWishListAsync()

**Review by Mikhail-Beloded — CHANGES_REQUESTED** [review]

**Review by Mikhail-Beloded — APPROVED** [review]

**Pull request merged**